### DISCUSSION TITLE
How can I run Halide on a GPU simulator instead of an actual GPU?

### DISCUSSION THREAD

**Discussion by LongPham7** [halide/Halide#5486] [Q&A]
I have been working on a research project where we investigate the impact of GPU's branch divergence in the setting of Halide. 
I don't have an actual GPU on my computer, but I instead have a GPU simulator called [GPGPU-Sim](https://github.com/gpgpu-sim/gpgpu-sim_distribution). Here are the details of my computing environment:
What I do NOT have on my computer:
1. NVIDIA GPU
2. NVIDIA GPU driver.

What I do have on my computer:
1. CUDA Toolkit version 11.0
2. GPGPU-Sim. 

I tried to JIT-compile, for instance, lesson_12_using_the_gpu.cpp from [Halide's tutorial](https://github.com/halide/Halide/blob/master/tutorial/lesson_12_using_the_gpu.cpp) using the command
``g++ lesson_12_using_the_gpu.cpp -g -std=c++11 -I <path/to/Halide.h> -I <path/to/tools/halide_image_io.h> -L <path/to/libHalide.so> -lHalide -L <path/to/cuda-11.0/bin/lib64> `libpng-config --cflags --ldflags` -ljpeg -lpthread -ldl -o lesson_12_using_the_gpu``,
where I additionally specify the path to cuda-11.0/bin/lib64. 
However, it failed to run. I suspect that it is because I don't have a GPU driver on my computer. 

Does anyone happen to know how to run Halide in such an environment? Thanks a lot in advance.

**Comment by zvookin** [comment]
I tried building GPGPU-Sim, but it isn't happening easily. My guess is you want to add the path to GPGPU-Sim's libcuda in the ```lib_names``` array in src/runtime/cuda.cpp. (Similarly for opencl if using it.) This involves building Halide from source, but my guess is you want to do this anyway. Or if you are using a binary build you can copy the code for the entire function ```halide_cuda_get_symbol``` change it to not be WEAK and add the pathname in your version.

I'm pretty sure it makes sense for us to provide a way to do this in Halide in general so a PR to get a pathname from an environment variable or an API to set the path will likely be a good thing. Patches welcome or at some point I'll finish getting GPGPU built and take care of it. This is assuming the approach I've suggested works. It is attractive to us as another alternative for testing things.

**Reply by LongPham7** [reply]
Thank you so much for your swift response!! I will try your suggestions. 

Installing GPGPU-Sim was a pain to me as well. Herewith I have enclosed my installation instructions of GPGPU-Sim: 
[Instructions_setting_up_GPGPU-Sim.pdf](https://github.com/halide/Halide/files/5611620/Instructions_setting_up_GPGPU-Sim.pdf). Hopefully, they work for you.
